File: src/main/java/io/bindernews/thegrackle/variables/AbstractSimpleVariable.java
package io.bindernews.thegrackle.variables;

import basemod.abstracts.DynamicVariable;
import com.evacipated.cardcrawl.modthespire.lib.SpireField;
import com.megacrit.cardcrawl.cards.AbstractCard;
import io.bindernews.bnsts.IVariable;
import lombok.*;

/**
 * This class splits the raw manipulation of a dynamic variable from its common
 * usage (e.g. {@link ExtraHitsVariable}), and reduces overall memory usage
 * by adding only one field to {@link AbstractCard} instead of three, and
 * only allocating instances as necessary.
 * <br/>
 * The field value should default to {@code null}. When one of the {@code set*}
 * methods of {@link AbstractSimpleVariable} is called, the internal field will be initialized.
 */
@AllArgsConstructor
public abstract class AbstractSimpleVariable extends DynamicVariable implements IVariable {
    /**
     * The field instance that we're storing values in.
     */
    protected final SpireField<VariableInst> field;
    public final VariableInst defaultValue;

    @Override
    public boolean isModified(AbstractCard card) {
        return value(card) != baseValue(card);
    }

    @Override
    public int value(AbstractCard card) {
        val f = field.get(card);
        return f == null ? defaultValue.value : f.value;
    }

    @Override
    public int baseValue(AbstractCard card) {
        val f = field.get(card);
        return f == null ? defaultValue.baseValue : f.baseValue;
    }

    @Override
    public boolean upgraded(AbstractCard card) {
        val f = field.get(card);
        return f == null ? defaultValue.upgraded : f.upgraded;
    }

    @Override
    public void setBaseValue(AbstractCard card, int amount) {
        val f = initField(card);
        f.value = amount;
        f.baseValue = amount;
        card.initializeDescription();
    }

    @Override
    public void setValue(AbstractCard card, int amount) {
        initField(card).value = amount;
    }

    @Override
    public void upgrade(AbstractCard card, int amount) {
        val f = initField(card);
        f.upgraded = true;
        f.baseValue += amount;
        f.value += amount;
    }

    /**
     * Returns the extraHits field of the card, initializing it if it's null.
     */
    private VariableInst initField(AbstractCard card) {
        var f = field.get(card);
        if (f == null) {
            f = new VariableInst(defaultValue.value, defaultValue.baseValue, defaultValue.upgraded);
            field.set(card, f);
        }
        return f;
    }

    /**
     * The backing field type that stores field data.
     */
    @EqualsAndHashCode
    @NoArgsConstructor
    @AllArgsConstructor
    public static final class VariableInst {
        public int value = -1;
        public int baseValue = -1;
        public boolean upgraded = false;
    }
}

File: src/main/kotlin/net/bindernews/grackle/variables/AbstractSimpleVariable.kt
package net.bindernews.grackle.variables

import basemod.abstracts.DynamicVariable
import com.evacipated.cardcrawl.modthespire.lib.SpireField
import com.megacrit.cardcrawl.cards.AbstractCard
import net.bindernews.grackle.helper.IVariable

/**
 * This class splits the raw manipulation of a dynamic variable from its common
 * usage (e.g. [ExtraHitsVariable]), and reduces overall memory usage
 * by adding only one field to [AbstractCard] instead of three, and
 * only allocating instances as necessary.
 * <br></br>
 * The field value should default to `null`. When one of the `set*`
 * methods of [AbstractSimpleVariable] is called, the internal field will be initialized.
 */
abstract class AbstractSimpleVariable(
    /** The field instance that we're storing values in. */
    @JvmField protected val field: SpireField<VariableInst?>,
    @JvmField val defaultValue: VariableInst,
) : DynamicVariable(), IVariable {
    override fun isModified(card: AbstractCard): Boolean {
        return value(card) != baseValue(card)
    }

    override fun value(card: AbstractCard): Int {
        val f = field[card]
        return f?.value ?: defaultValue.value
    }

    override fun baseValue(card: AbstractCard): Int {
        val f = field[card]
        return f?.baseValue ?: defaultValue.baseValue
    }

    override fun upgraded(card: AbstractCard): Boolean {
        val f = field[card]
        return f?.upgraded ?: defaultValue.upgraded
    }

    override fun setBaseValue(card: AbstractCard, amount: Int) {
        val f = initField(card)
        f.value = amount
        f.baseValue = amount
        card.initializeDescription()
    }

    override fun setValue(card: AbstractCard, amount: Int) {
        initField(card).value = amount
    }

    override fun upgrade(card: AbstractCard, amount: Int) {
        val f = initField(card)
        f.upgraded = true
        f.baseValue += amount
        f.value += amount
    }

    /**
     * Returns the [VariableInst] field of the card, initializing it if it's null.
     */
    private fun initField(card: AbstractCard): VariableInst {
        var f = field[card]
        if (f == null) {
            f = VariableInst(defaultValue.value, defaultValue.baseValue, defaultValue.upgraded)
            field[card] = f
        }
        return f
    }

    /**
     * The backing field type that stores field data.
     */
    data class VariableInst(var value: Int = -1, var baseValue: Int = -1, var upgraded: Boolean = false)
}